### PR TITLE
Allow BelongsTo relations to be created on item creation when using relationship name as field name

### PR DESCRIPTION
**Pull request by mgralikowski**
Main context:
https://github.com/Laravel-Backpack/CRUD/issues/1568
Also:
https://github.com/Laravel-Backpack/CRUD/issues/2826

@tabacitu mentioned here:
https://github.com/Laravel-Backpack/CRUD/issues/3011#issuecomment-655917152
that there are some changes in relationship logic in 4.x. I tried and I agree, now it is possible to save relationships directly, without making foreign key fillable (in my opinion still is not the best practice to do it :).  But I see one problem.

Example with simple 1:1 relationship to user:
```php
$this->crud->addField([
    'name'  => 'user',
    'label' => 'User',
    'type'  => 'relationship',
]);
```
If user_id column is nullable, I do not have any problem with saving a new entity because in the first step we are saving the model without the user, and in the next step will be associated with the user model. But if user_id is not nullable it will crash because:
`General error: 1364 Field 'user_id' doesn't have a default value`
This PR change only the moment where we saving the final model.

What do you think?

### CODE REVIEW

**Comment by tabacitu** [comment]
Wow... just... wow... it's that simple a fix @mgralikowski ? Just use `make()`&`save()` instead of `create()`? Then definitely, I love it. I totally agree this is a whole we need to plug ASAP.

I bet it took quite a lot of debugging and trial&error to get the PR this small - thanks a lot!

**Comment by tabacitu** [comment]
Just tried this in NewsCRUD and... I seem to get exactly the opposite of the expected result 🤣 `master` works, `mgralikowski-patch-12` does not... Am I doing something wrong here? Do I not understand the problem being addressed?

Here's what I'm doing. I'm adding an `Article` with a related `Category` (1-n relationship, non nullable) and a few `Tags` (n-n relationship, nullable):

Screenshot 2020-09-12 at 08 21 19

On `master` it work fine, added the entry correctly.

On your branch, `mgralikowski-patch-12`, it didn't work at all, it threw an error because of the n-n relationship:

Screenshot 2020-09-12 at 08 21 02


So, to sum up:
- `master`
   - ✅ 1-n non-nullable relationship saved fine
   - ✅ n-n relationship saved fine
- `mgralikowski-patch-12`
   - ✅ 1-n non-nullable relationship saved fine
   - 🟥 n-n relationship did NOT save fine - it threw an error

Which is exactly the opposite of what's reported in #2826 , right?! Any idea how I'm testing this wrong @mgralikowski ?

Note: In terms of #1568, `category_id` was fillable in both tests. I've also tried it _without_ a fillable `category_id` and it has the same results as above. So `fillable` does not seem to influence the results one way or the other.

**Comment by tabacitu** [comment]
Update: see https://github.com/Laravel-Backpack/CRUD/issues/2826#issuecomment-691483348 - I've tried with the fluent syntax too, thinking that might be a problem, but no, same results.

**Comment by mgralikowski** [comment]
Ok, so fix will be not so easy.

This is 100% sure that this line:
`$item = $this->model->create(Arr::except($data, $nn_relationships));`
https://github.com/Laravel-Backpack/CRUD/blob/2c87086173c0c1b746d6f0fd49783f6eefa787e9/src/app/Library/CrudPanel/Traits/Create.php#L31
will throw error when we try to create and **store** model without required foreign key. It is basics of Eloquent. But this problem is associated only with **BelongsTo** (1:1) relation.

I tried the news library and the reason, why there is no error, is:
- a `category` is normal field, type `select`, no `relationship`
- a `category_id` is made as fillable in model
When you remove `category_id` from fillable it will cause an error.

I will back with a more advanced solution.

**Comment by mgralikowski** [comment]
Ok. Done. It is fast commit, of course I would finnaly move logic to separate methods.
In this commit I moved BelongTo handling before save and thats all. 

If You want to test is with news library you need make here 2 changes:

- Remove `category_id` from fillable in model
- Change category into relationship type so it will look like this:
```php
            $this->crud->addColumn([
                'label' => 'Category',
                'type' => 'relationship',
                'name' => 'category',
                'entity' => 'category',
                'attribute' => 'name',
                'wrapper'   => [
                    'href' => function ($crud, $column, $entry, $related_key) {
                        return backpack_url('category/'.$related_key.'/show');
                    },
                ],
            ]);
```

**Comment by mgralikowski** [comment]
I commited a new version, more readable and optimal.

**Comment by tabacitu** [comment]
Thanks a lot for this @mgralikowski . 

@pxpm what do you think?

**Comment by pxpm** [comment]
@tabacitu @mgralikowski 

Hello guys. From my perspective I think this is a something we should add.

What this does is that `make()` creates a model instance without saving to database, attaches the belongs to relations (Like if the user have specified a field `article_id` and added it into the form) so it will be in the create request. 

https://recordit.co/UEz2kJvKKR

I think this change will make the app behave the same when you add the field using only relationship name, or providing the full column as a field to be in request.

About the events, I am concerned that this will not provide `creating/created` and only `saving/saved`. Do you happen to know @mgralikowski ?

Best,
Pedro

**Comment by mgralikowski** [comment]
Hm, I added to the _Article_ model this:

```php
    protected static function boot()
    {
        parent::boot();
        static::creating(static function (self $article) {
            dd($article);
        });
    }
```
And _dd_ is executing normally:
obraz

**Comment by mgralikowski** [comment]
Anyway, this PR is not finished yet. I forgot about the update action. Ok, for now, there is no error, but we can't change the category.

**Comment by mgralikowski** [comment]
Bugfix, refactor and handling update action as well.

**Comment by pxpm** [comment]
So testing this solution it works great. Except for the updating/updated events. 

I added some code in update() function to manually trigger the events:

```php
....

$item = $this->handleBelongsToRelations($item, $data);
        $item->fill($data);
        event('eloquent.updating:'.$this->model, $item);
        $item->save();
        event('eloquent.updated:'.$this->model, $item);
```

Would you mind creating a branch in backpack/crud repository instead of using your repository, so we can push any fix to it and you still be the contributor ?

This are the steps that I would use to create a branch in crud repository:

1 - delete the vendor/backpack/crud folder.
2 - `composer install --prefer-source`
3 - navigate to vendor/backpack/crud and `git checkout master`, `git pull`
4 - create your branch: `git checkout -b my-branch-name`
5 - add your code.
6 - `git commit -am "comment for the fix"`
7 - `git push origin my-branch-name`
8 - Navigate to backpack/crud in browser, you will have a yellow box that will allow you to create the PR from the branch you submited. 

This can be done using git GUI. Note that if you have customized vendor folder (you shouldn't) all changes will be lost.

This way me and @tabacitu could push any fixes, even merge master to test with the latest changes. 

Thanks for this, I totally agree with this solution. 

Best to you,
Pedro

**Comment by mgralikowski** [comment]
@pxpm I tried this instructions but I fail on:
```
/vendor/backpack/crud (relationships-refactor-logic)
$ git push origin relationships-refactor-logic
ERROR: Permission to Laravel-Backpack/CRUD.git denied to mgralikowski.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
I am not sure how to create branch without access.
I created new PR with branch, but still on my fork: https://github.com/Laravel-Backpack/CRUD/pull/3316

**Comment by pxpm** [comment]
Closing this in favor of: #3250
